src/main/java/org/opensearch/knn/quantization/models/quantizationState/QuantizationStateCache.java

```diff
@@ -0,0 +1,132 @@
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+
+import static org.opensearch.knn.index.KNNSettings.QUANTIZATION_STATE_CACHE_EXPIRY_TIME_MINUTES_SETTING;
+import static org.opensearch.knn.index.KNNSettings.QUANTIZATION_STATE_CACHE_SIZE_LIMIT_SETTING;
+
+/**
+ * A thread-safe singleton cache that contains quantization states.
+ */
+@Log4j2
+public class QuantizationStateCache {
+
+    private static volatile QuantizationStateCache instance;
+    private static ClusterService clusterService;
+    private Cache<String, QuantizationState> cache;
+    private long maxCacheSizeInKB;
+    private Instant evictedDueToSizeAt;
+
+    private QuantizationStateCache() {
+        maxCacheSizeInKB = QUANTIZATION_STATE_CACHE_SIZE_LIMIT_SETTING.get(clusterService.getSettings()).getKb();
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(QUANTIZATION_STATE_CACHE_SIZE_LIMIT_SETTING, it -> {
+            maxCacheSizeInKB = it.getKb();
+            rebuildCache();
+        });
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(QUANTIZATION_STATE_CACHE_EXPIRY_TIME_MINUTES_SETTING, it -> {
+            rebuildCache();
+        });
+        buildCache();
+    }
+
+    /**
+     * Gets the singleton instance of the cache.
+     * @return QuantizationStateCache
+     */
+    public static QuantizationStateCache getInstance() {
+        if (instance == null) {
+            synchronized (QuantizationStateCache.class) {
+                if (instance == null) {
+                    instance = new QuantizationStateCache();
+                }
+            }
+        }
+        return instance;
+    }
+
+    /**
+     * Initialize the cache
+     * @param clusterService used to update settings
+     */
+    public static void initialize(ClusterService clusterService) {
+        QuantizationStateCache.clusterService = clusterService;
+    }
+
+    private void buildCache() {
+        this.cache = CacheBuilder.newBuilder().concurrencyLevel(1).maximumWeight(maxCacheSizeInKB).weigher((k, v) -> {
+            try {
+                return ((QuantizationState) v).toByteArray().length;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        })
+            .expireAfterAccess(
+                QUANTIZATION_STATE_CACHE_EXPIRY_TIME_MINUTES_SETTING.get(clusterService.getSettings()).getMinutes(),
+                TimeUnit.MINUTES
+            )
+            .removalListener(this::onRemoval)
+            .build();
+    }
+
+    private void rebuildCache() {
+        clear();
+        buildCache();
+    }
+
+    /**
+     * Retrieves the quantization state associated with a given field name.
+     * @param fieldName The name of the field.
+     * @return The associated QuantizationState, or null if not present.
+     */
+    public QuantizationState getQuantizationState(String fieldName) {
+        return cache.getIfPresent(fieldName);
+    }
```

Comment:
Shouldn't this be more of a logic where if entry is not present in the cache then we load the value in the cache and then return that value?

Any reason of not making this as Read-Through cache?

@Vikasht34 thoghts?

Reply:
Well here is my thoughts on this , I am really not in favour of clubbing cache with Quantization Reader, since we don't know what would be future for Cache interms of unifying intefaces across multiple caches and implementation , I would keep this totally light weight, 

Neither I want my NativeQuantizationState Reader to know about cache .

QuantizationStateCache: Manages in-memory caching of quantization states.
QuantizationStateReader: Handles reading quantization states from persistent storage.
QuantizationStateWriter: Manages writing quantization states to persistent storage.

From Query flow 
I will have my Interface 

```
public interface QuantizationStateLoader {
    QuantizationState getQuantizationState(String fieldName, String typeofQuantization);
}
```

and this inteface will handle both cache and NativeQuantizationStateReader , to read it from cache , if not read from file , deseralize based on type of Quantization and save it Cache. 


So my Cache is clean , neither it has to know about type of Quantization State to deseralize and neither about how it was originally stored.

Reply:
> So Cache is clean , neither it has to know about type of Quantization State to deseralize and neither about how it was originally stored.

if are looking for making cache clean then the current implementation is not clean. It already has knowledge of `QuantizationState` object and other QS details. 

But whether we want to make cache clean or not that is separate question. But it doesn't ans the question why this cache is not a Read-Through cache? You can just provide a loader functional interface here which is just loads the entry in cache if its not present. Then no-one needs to know about QS.

Reply:
The problem with this kind of design is, we will have to write a wrapper on top of it which will first check if the entry in cache is present or not, and then load data in the cache. Which can be just avoided here.